File: role.py
from flask import Blueprint,request,jsonify
from models import Role

roleBp=Blueprint("roleBp",__name__)

@roleBp.post("/role/new")
def newRole():
    try:
        role=request.get_json()

        if not role:
            return jsonify({"status":"error","message":"Role Not To Be Empty."}), 400
        
        name=role.get("name")

        if not name:
            return jsonify({
                "status": "error",
                "message": "Role name is required."
            }), 400
        
        if Role.objects(name=name).first():
            return jsonify({
                "status": "error",
                "message": "Role already exists."
            }), 409
        
        Role(
            name=name
        ).save()

        return jsonify({"status": "success", "message": "Role Added Successfully."}), 201
    
    except Exception as e:
        return jsonify({"status":"error","message":f"Error {str(e)}"}), 500
    
@roleBp.get("/role/getAll")
def allRoles():
    try:
        roles=Role.objects()

        if not roles:
            return jsonify({"status":"error","message":"Roles Empty."}), 200
        
        roleList=[]

        for role in roles:
            data={
                "name":role.name
            }

            roleList.append(data)

        return jsonify({"status":"success","message":"Roles Are Retrieved Successfully.","data":roleList}), 200

    except Exception as e:
        return jsonify({"status":"error","message":f"Error {str(e)}"}), 500
    
@roleBp.get("/role/getSpecific")
def roleSpecific():
     try:
        id = request.args.get("id")

        if not id:
            return jsonify({"status": "error", "message": "Id is required."}), 400 
        
        role=Role.objects(id=id).first()

        if not role:
            return jsonify({"status":"error","message":"Role Not Found."}), 404 
        
        data={
            "name":role.name
            }

        return jsonify({"status":"success","message":"Role Retrieved Successfully.","data":data}), 200
     
     except Exception as e:
        return jsonify({"status":"error","message":f"Error {str(e)}"}), 500 
     
@roleBp.post("/role/update")
def roleUpdate():
    try:
        id = request.args.get("id")

        if not id:
            return jsonify({"status": "error", "message": "Id is required."}), 400     
        
        roleName=request.get_json()

        if not roleName:
            return jsonify({"status":"error","message":"Role Not To Be Empty."}), 400  
        
        name=roleName.get("name")

        if not name:
            return jsonify({
                "status": "error",
                "message": "Role name is required."
            }), 400

        if Role.objects(name=name).first():
            return jsonify({"status":"error","message":"Duplicated Role."}), 409
        
        role=Role.objects(id=id).first()

        if not role:
            return jsonify({"status":"error","message":"Role Not Found."}), 404 

        role.name=name

        role.save()

        return jsonify({"status":"success","message":"Role Updated Successfully."}), 200 
    
    except Exception as e:
        return jsonify({"status":"error","message":f"Error {str(e)}"}), 500
    
@roleBp.delete("/role/delete")
def roleDelete():
    try:
        id = request.args.get("id")

        if not id:
            return jsonify({"status": "error", "message": "Id is required."}), 400  

        role=Role.objects(id=id).first()

        if not role:
            return jsonify({"status":"error","message":"Role Not Found."}), 404 
        
        role.delete()

        return jsonify({"status":"success","messge":"Role Deleted Successfully."}), 200 
    
    except Exception as e:
        return jsonify({"status":"error","message":f"Error {str(e)}"}), 500

File: static/js/role.js
async function fetchJSON(url, options = {}) {
    const res = await fetch(url, options);
    const data = await res.json();

    if (data.status !== "success") {
        throw new Error(data.message);
    }
    return data;
}

function formatDate(dt) {
    if (!dt) return "N/A";
    const d = new Date(dt);
    return d.toLocaleDateString("en-Us", {
        day: "2-digit",
        month: "short",
        year: "numeric",
        hour:"2-digit",
        minute:"2-digit",
        timeZone: "Asia/Kolkata"
    });
}

async function loadRoles() {
    try {
        const res = await fetchJSON("/role/getAll");
        const tbody = $("table tbody");
        tbody.empty();

        let sNo = 1;

        res.data.forEach(role => {
            tbody.append(`
                <tr data-id="${role.id || ''}">
                    <td class="sNo">${sNo++}</td>
                    <td class="role">${role.name}</td>
                    <td class="createdAt">${formatDate(role.createdAt)}</td>
                    <td class="updatedAt">${formatDate(role.updatedAt)}</td>
                    <td class="text-center">
                        <div class="dropdown">
                            <button class="btn btn-sm btn-outline-secondary" data-bs-toggle="dropdown">
                                <i class="bi bi-three-dots-vertical"></i>
                            </button>

                            <ul class="dropdown-menu">
                                <li>
                                    <button class="dropdown-item editRoleBtn">
                                        <i class="bi bi-pencil me-2"></i>Edit
                                    </button>
                                </li>
                                <li>
                                    <button class="dropdown-item text-danger deleteRoleBtn">
                                        <i class="bi bi-trash me-2"></i>Delete
                                    </button>
                                </li>
                            </ul>
                        </div>
                    </td>
                </tr>
            `);
        });

        if (sNo === 1) {
            tbody.append(`
                <tr>
                    <td colspan="5" class="text-center text-muted">
                        No roles found
                    </td>
                </tr>
            `);
        }

    } catch (err) {
        alert(err.message);
    }
}


$("#createRoleBtn").on("click", async function () {
    const roleName = $("#roleName").val().trim();

    if (!roleName) {
        alert("Role name is required");
        return;
    }

    try {
        await fetchJSON("/role/new", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ name: roleName })
        });

        $("#roleName").val("");
        $("#newRoleModal").modal("hide");

        alert("Role created successfully");
        loadRoles();

    } catch (err) {
        alert(err.message);
    }
});


$(document).on("click", ".editRoleBtn", function () {
    const row = $(this).closest("tr");
    const roleName = row.find(".role").text();

    const newName = prompt("Edit role name", roleName);
    if (!newName || newName.trim() === roleName) return;

    updateRole(row.data("id"), newName.trim());
});

async function updateRole(roleId, name) {
    try {
        await fetchJSON(`/role/update?id=${roleId}`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ name })
        });

        alert("Role updated successfully");
        loadRoles();

    } catch (err) {
        alert(err.message);
    }
}



$(document).on("click", ".deleteRoleBtn", function () {
    const row = $(this).closest("tr");
    const roleName = row.find(".role").text();
    const roleId = row.data("id");

    if (!confirm(`Delete role "${roleName}"?`)) return;

    deleteRole(roleId);
});

async function deleteRole(roleId) {
    try {
        await fetchJSON(`/role/delete?id=${roleId}`, {
            method: "DELETE"
        });

        alert("Role deleted successfully");
        loadRoles();

    } catch (err) {
        alert(err.message);
    }
}

$(document).ready(loadRoles);
